### PR TITLE
[WIP] Add information about names of UI element to docs

### LINKED ISSUE #906: Add a user interface visual to docs which labels UI items
A nice documentation enhancement in the notebook documentation would be to add a "what's it called" visual for the User interface items. Reference https://github.com/jupyter/enhancement-proposals/issues/9#issuecomment-167805546

### PR DESCRIPTION
**Pull request by captainsafia**
Closes #906.

Opening this up because I have an issue with building the docs. When I run `make html` I get the following:

```
jupyter nbconvert --to rst source/examples/Notebook/*.ipynb --template=source/template --FilesWriter.build_directory=source/examples/Notebook/rstversions
/Users/safia/dev/ipython/IPython/nbconvert.py:13: ShimWarning: The `IPython.nbconvert` package has been deprecated. You should import from nbconvert instead.
  "You should import from nbconvert instead.", ShimWarning)
[NbConvertApp] Converting notebook source/examples/Notebook/Connecting with the Qt Console.ipynb to rst
Traceback (most recent call last):
  File "/Users/safia/anaconda3/bin/jupyter-nbconvert", line 9, in <module>
    load_entry_point('nbconvert==4.2.0.dev0', 'console_scripts', 'jupyter-nbconvert')()
  File "/Users/safia/anaconda3/lib/python3.5/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/Users/safia/anaconda3/lib/python3.5/site-packages/traitlets/config/application.py", line 578, in launch_instance
    app.start()
  File "/Users/safia/dev/nbconvert/nbconvert/nbconvertapp.py", line 289, in start
    self.convert_notebooks()
  File "/Users/safia/dev/nbconvert/nbconvert/nbconvertapp.py", line 412, in convert_notebooks
    self.convert_single_notebook(notebook_filename)
  File "/Users/safia/dev/nbconvert/nbconvert/nbconvertapp.py", line 383, in convert_single_notebook
    output, resources = self.export_single_notebook(notebook_filename, resources)
  File "/Users/safia/dev/nbconvert/nbconvert/nbconvertapp.py", line 335, in export_single_notebook
    output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
  File "/Users/safia/dev/nbconvert/nbconvert/exporters/exporter.py", line 165, in from_filename
    return self.from_notebook_node(nbformat.read(f, as_version=4), resources=resources, **kw)
  File "/Users/safia/dev/nbconvert/nbconvert/exporters/templateexporter.py", line 204, in from_notebook_node
    raise IOError('template file "%s" could not be found' % self.template_file)
OSError: template file "source/template" could not be found
make: *** [ipynb2rst] Error 1
```

I tried updating the `template` variable to `source/template.tpl` (which made sense to me) as well as `./source/template` (which also made sense to me), to no avail. What am I missing?

Below is information about my setup.

```
safia at eniac in docs $ pip show sphinx
---
Metadata-Version: 1.1
Name: Sphinx
Version: 1.3.1
Summary: Python documentation generator
Home-page: http://sphinx-doc.org/
Author: Georg Brandl
Author-email: georg@python.org
License: BSD
Location: /Users/safia/anaconda3/lib/python3.5/site-packages/Sphinx-1.3.1-py3.5.egg
Requires: six, Jinja2, Pygments, docutils, snowballstemmer, babel, alabaster, sphinx-rtd-theme
Entry-points:
  [console_scripts]
  sphinx-apidoc = sphinx.apidoc:main
  sphinx-autogen = sphinx.ext.autosummary.generate:main
  sphinx-build = sphinx:main
  sphinx-quickstart = sphinx.quickstart:main
  [distutils.commands]
  build_sphinx = sphinx.setup_command:BuildDoc
```

```
safia at eniac in docs $ jupyter nbconvert --version
4.2.0.dev
```
cc: @willingc

### CODE REVIEW

**Comment by willingc** [comment]
@captainsafia Graphic looks good. Thanks for doing that :+1:

On the build (sorry if I'm repeating what you already know), the `Makefile` does three major steps when you enter `make html`:

Step 1 - Autogenerates a config.rst file that covers the different configuration options available [tagged: source/config.rst]

Step 2 - Convert the ipynb files in `docs/source/Notebook/examples` to rst files and place in `rstversions` directory using nbconvert [tagged: ipynb2rst in the Makefile]

Step 3 - Build the sphinx documentation of all the rst files

Snippet from Makefile

```
html: source/config.rst ipynb2rst
	$(SPHINXBUILD) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
	@echo
	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."

source/config.rst:
	python3 autogen_config.py
	@echo "Created docs for config options"

ipynb2rst:
	jupyter nbconvert --to rst source/examples/Notebook/*.ipynb --template=source/template --FilesWriter.build_directory=source/examples/Notebook/rstversions
	@echo "Converted notebooks to rst"
```

I looked in my virtualenv that I use for docs and it has nbconvert v4.1.0 and something could have broken between that and the 4.2.0dev version.

I was playing around with nbconvert for ipynb to rst and md files earlier today. I will let you know if I find the root cause here.

You might wish to try if you haven't already:
```
make clean
make html
```
make clean will delete the build output directories so you will make sure Sphinx isn't leaving any cruft between builds.

**Comment by willingc** [comment]
Two gists for @captainsafia 
[requirements.txt](https://gist.github.com/willingc/c42ebf9ad43386f1d5f8) used for doc development

After ``pip install -r requirements.txt`` in my virtualenv with 3.5.1, [this gist](https://gist.github.com/willingc/8c41ffd7b2bfb587e41d) has the output of ``pip freeze`` which includes dependencies.

**Comment by willingc** [comment]
@captainsafia I was able to build without error with the latest nbconvert. 

Hmm...I wonder what the delta is between the anaconda config that you have and my virtualenv. Can you compare what's in your environment to the output gist linked above? Perhaps it is ``path.py`` or ``traitlets`` that is different in yours. Yay! We :heart_eyes: computers. :speak_no_evil:

**Comment by captainsafia** [comment]
Hm. So here is what I did to resolve the issue.

1. Set up a new conda environment with Python 3.5.1.
2. Installed `jupyter` and other basic requirements.
3. Installed from the `doc-requirements.txt` you gave me.
4. Ran `pip freeze > my-output.txt`.
5. Ran `diff output.txt my-output.txt` where `output.txt` is the output of your `pip freeze`.
6. Confirmed that we had the same versions for things (except `ipython-master` where I am on a commit ahead of yours I assume because I installed later).
7. Ran `make clean && make html` again and the same issue arose.

It looks like we also have the same `Makefile` in place. I'll tinker more and see what I get.

**Comment by willingc** [comment]
@captainsafia I need a theme song. On a serious note, I tried to build the docs on a different OSX system of mine that had anaconda installed. Here is the lovely output which kindly tells me that pandoc is not found.

So, I then tried building with a virtualenv with pip and docs built with no issues.

Completely deleted anaconda. Downloaded and installed it. Tried building in a configured conda environment. Same error about pandoc. Well, isn't that special.

Exited conda environment. Uninstalled brew and reinstalled brew and then installed pandoc using brew. Docs still built with the virtualenv with pip. Reactivated the conda environment and then the conda env built the docs cleanly.

Wiser minds may be able to explain why.

```
(docsafia)Carols-Mac-mini:docs carol1$ make html
jupyter nbconvert --to rst source/examples/Notebook/*.ipynb --template=source/template --FilesWriter.build_directory=source/examples/Notebook/rstversions
[NbConvertApp] Converting notebook source/examples/Notebook/Connecting with the Qt Console.ipynb to rst
[NbConvertApp] ERROR | Error while converting 'source/examples/Notebook/Connecting with the Qt Console.ipynb'
Traceback (most recent call last):
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/nbconvertapp.py", line 335, in export_single_notebook
    output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/exporter.py", line 165, in from_filename
    return self.from_notebook_node(nbformat.read(f, as_version=4), resources=resources, **kw)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/templateexporter.py", line 202, in from_notebook_node
    output = self.template.render(nb=nb_copy, resources=resources)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./source/template.tpl", line 5, in top-level template code
    `View the original notebook on nbviewer <http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/{{ resources['metadata']['path'] }}/{{ resources['metadata']['name'] | replace(' ', '%20') }}.ipynb>`__
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/../templates/rst.tpl", line 1, in top-level template code
    {%- extends 'display_priority.tpl' -%}
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/../templates/skeleton/display_priority.tpl", line 1, in top-level template code
    {%- extends 'null.tpl' -%}
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/../templates/skeleton/null.tpl", line 26, in top-level template code
    {%- block body -%}
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/../templates/skeleton/null.tpl", line 28, in block "body"
    {%- block any_cell scoped -%}
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/exporters/../templates/skeleton/null.tpl", line 72, in block "any_cell"
    {%- block markdowncell scoped-%}
  File "./source/template.tpl", line 19, in block "markdowncell"
    {{ cell.source | markdown2rst | replace(".ipynb>", ".html>") }}
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/filters/markdown.py", line 87, in markdown2rst
    return pandoc(source, 'markdown', 'rst', extra_args=extra_args)
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/utils/pandoc.py", line 49, in pandoc
    check_pandoc_version()
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/utils/pandoc.py", line 95, in check_pandoc_version
    v = get_pandoc_version()
  File "/Users/carol1/anaconda/envs/docsafia/lib/python3.5/site-packages/nbconvert/utils/pandoc.py", line 74, in get_pandoc_version
    raise PandocMissing()
nbconvert.utils.pandoc.PandocMissing: Pandoc wasn't found.
Please check that pandoc is installed:
http://johnmacfarlane.net/pandoc/installing.html
make: *** [ipynb2rst] Error 1
```

**Comment by willingc** [comment]
@captainsafia Let me know what next steps seem reasonable on your system. Anything from A) install Python 3.5 from Python.org, create a virtualenv and pip install stuff in the activated virtualenv to B) Uninstall and reinstall anaconda (which has solved issues on django projects when using conda) to C) call Ghostbusters :ghost: are reasonable choices.

**Comment by captainsafia** [comment]
So I uninstalled and reinstalled anaconda, created the new environment, installed the dependencies, and......didn't work.

So I uninstalled anaconda, setup virtualenv with the Python3 on my system, installed all the dependencies, and....didn't work.

So that rules out this being an Anaconda specific error. I'm going to declare this as an `nbconvert` error. I looked through some old issues on `nbconvert` and it looks like I'm not the first person to get this error in a situation where the template file actually did exist. I'll explore this path and report back...

Good thing I read a lot of Nancy Drew when I was little. :mag:

**Comment by willingc** [comment]
@captainsafia Your detective powers are excellent. Which OS are you using? I'm assuming a Linux flavor.

**Comment by captainsafia** [comment]
I'm on OS X. It appears that my specific issue is relevant to what was reported in jupyter/nbconvert#163 about a month ago. In which case, the error about a template not being found is actually raised if there is an error *within* the template code (frustrating). Is your template.tpl file different from the one in the repository in any way?

**Comment by willingc** [comment]
Whoa. I wouldn't have guessed OS X. I'm pretty sure that the template code that I have is from just "git clone repo". Very odd. I wonder if for some reason under certain conditions, pandoc is being hidden/not found and erroring with something not completely helpful. I did get a pretty prominent "pandoc not found" on one of my anaconda tries. For curiousity send details re: OS X version, retina/non-retina display, favorite flavor of apple pie... I'll look at nbconvert too.

**Comment by captainsafia** [comment]
I've got `pandoc` already installed. Also, when I removed the `--template` flag from the command in the Makefile and ran `make clean && make html` again, everything was built properly which is why I think this is specifically an issue with nbconvert and templates. Here is any and all information that might be relevant in this specific pursuit.

Device: Macbook Air
OS: OS X El Capitan (10.11.12)
Favorite Flavor of Pie: Cinnamon
Display: Non-Retina
Horoscope: Leo the Lion
`pandoc` version: 1.15.2.1

**Comment by willingc** [comment]
Excellent detective work @captainsafia! I think that you are quite right on the direction to pursue. (Only real difference that I see is that you are on El Capitan and I am still on Yosemite.) I will give it a try on Windows 10 tomorrow just for fun :-) Nice work.

**Comment by captainsafia** [comment]
Yeah, in the meanwhile, I'll remove the `--template` flag on my local (doesn't seem to be relevant to what I am doing now so...) and continue working on the UI docs for this PR.

**Comment by captainsafia** [comment]
Alright, I've pushed some changes to this branch.

Should I add more detail or is this sufficient?

File: docs/source/ui_components.rst [hunk 1]
```diff
@@ -0,0 +1,32 @@
+UI Components
+=============
+When opening bug reports or sending emails to the Jupyter mailing list, it is
+useful to know the names of different UI component so that other developers
```
Comment (willingc) [inline docs/source/ui_components.rst]: components

File: docs/source/ui_components.rst [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+and users have an easier time helping you diagnose your problems. This section
+of the documentation is designed to familiarize you with the different names
+of UI elements and the different states that UI elements can be in.
```
Comment (willingc) [inline docs/source/ui_components.rst]: Let's tighten this up.

Perhaps: This section will familiarize you the names of UI elements and the different Notebook modes (i.e. editing, command mode).

File: docs/source/ui_components.rst [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
```
Comment (willingc) [inline docs/source/ui_components.rst]: Perhaps subheading "Notebook Dashboard"

File: docs/source/ui_components.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+When you launch ``jupyter notebook`` the first page that you encounter is the
+Notebook Dashboard.
+
+.. image:: /_static/images/jupyter-notebook-dashboard.png
+
+Once you've selected a Notebook to edit, the Notebook will open in the Notebook
```
Comment (willingc) [inline docs/source/ui_components.rst]: Subheading: Notebook Editor

File: docs/source/ui_components.rst [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+Editor.
+
+.. image:: /_static/images/jupyter-notebook-default.png
+
+If you would like to learn more about the specific elements within the Notebook
```
Comment (willingc) [inline docs/source/ui_components.rst]: Subheading: Interactive User Interface Tour of the Notebook

File: docs/source/ui_components.rst [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+Editor, you can go through the User Interface Tour by selecting Help in the
+menubar then selecting User Interface Tour.
+
+When a cell is in edit mode, the Cell Mode Indicator will change to reflect
```
Comment (willingc) [inline docs/source/ui_components.rst]: Perhaps add a brief mode indicator subsection mentioning what the mode indicators are.

Comment (willingc) [inline docs/source/ui_components.rst]: Subsection: Edit mode and Notebook Editor

File: docs/source/ui_components.rst [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+the cell's state.
+
+.. image:: /_static/images/jupyter-notebook-edit.png
+
+Now let's say that you've chosen to open a Markdown file instead of a Notebook
+file whilst in the Notebook Dasboard. If so, the file will be opened in the
```
Comment (willingc) [inline docs/source/ui_components.rst]: typo Dashboard

**Comment by willingc** [comment]
@captainsafia Sorry for the review delay. Overall, I like it. I have some tweaks that will likely make it more user friendly (mostly adding some subheadings).

Also, let's call out somewhere (perhaps in the graphic and text) that the Cell Mode indicator is essentially no indicator or blank. When it enters edit mode, the pencil icon is displayed.

Looking good :cat:

**Comment by captainsafia** [comment]
Darn, I gotta snap out of it with that spelling though.

I've made the requested updates.

**Comment by willingc** [comment]
@captainsafia Nicely done. I'm happy to merge. :sunflower:

**Comment by captainsafia** [comment]
I'm writing this comment to describe how I resolved the issue described above should any other developers come across it while Googling. The discovery of the solution was a humbling reminder of the folly of man and the inevitable rise of our machine overlords.

Upon running the command, the machine humbly warns me.

```
[NbConvertApp] WARNING | Collisions detected in jupyter_nbconvert_config.py and jupyter_nbconvert_config.json config files. jupyter_nbconvert_config.json has higher priority: {
  "Exporter": {
    "template_path": "['.', '/Users/paulperry/Library/Jupyter/templates'] ignored, using [u'/Users/paulperry/Library/Jupyter/templates']"
  }
}
```

Suffering from ego, I heeded this warning not and continued forth with my fruitless journey. Days later I returned to the challenge. This time I was wiser and read the machine's cryptic scriptures more closely. Behold! The machine informs me of the error. At which point, I accompany my binary-based companion in a friendly game of Unix-ing.

```
cd ~/.jupyter
vim jupyter_nbconvert_config.json
```

And added ".", the current working directory under `template_paths`. I ran the command again, and behold it worked! At this point, I turned off my machine, took a long walk, and pondered the fate of Man on this Earth.

Computer: 1. Safia: 0.

**Pull request merged**